library/core/src/ptr/non_null.rs

```diff
@@ -812,6 +821,11 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[stable(feature = "non_null_convenience", since = "1.80.0")]
     #[rustc_const_stable(feature = "non_null_convenience", since = "1.80.0")]
+    #[requires(kani::mem::same_allocation(self.as_ptr() as *const(), origin.as_ptr() as *const()))]
```

Comment:
This should also be safe if the pointers hold the same address, even if they are dangling

Reply:
Please see my question above.